UCLA QCBio RNAseq1 workshop — Homework
==========

Nicolas Rochette, February 2021.

Homework
----------

Focussing on the RNAseq methodology of either of the two articles listed below
(whichever your prefer, and you *need not* read them completely), answer the
following question in **a couple sentences each**:

1. What tissue was RNA extracted from?
2. What RNA fractions were selected for sequencing (i.e. did the authors use
   polyA-tail capture to isolate mRNAs, or did they use rRNA depletion) ?
3. How was the sequencing performed?
4. What software was used to align reads, to what reference genome?
5. What software was used to quantify gene expression from read alignments?
6. Briefly, what question(s) do the authors answer using RNAseq (i.e. gene
   expression quantifications) ?

Please send your answers to nrochette@ucla.edu by
Thursday, May 27, 2021.

Articles:

Kim, J.-W., Yang, H.-J., Brooks, M. J., Zelinger, L., KarakÃ¼lah, G., Gotoh, N.,
Boleda, A., Gieser, L., Giuste, F., Whitaker, D. T., Walton, A., Villasmil, R.,
Barb, J. J., Munson, P. J., Kaya, K. D., Chaitankar, V., Cogliati, T., &
Swaroop, A. (2016). NRL-Regulated Transcriptome Dynamics of Developing Rod
Photoreceptors. Cell Reports, 17(9), 2460-2473.
https://doi.org/10.1016/j.celrep.2016.10.074

or

van Schouwenburg, P. A., Davenport, E. E., Kienzler, A.-K., Marwah, I., Wright,
B., Lucas, M., Malinauskas, T., Martin, H. C., WGS500 Consortium, Lockstone, H.
E., Cazier, J.-B., Chapel, H. M., Knight, J. C., & Patel, S. Y. (2015).
Application of whole genome and RNA sequencing to investigate the genomic
landscape of common variable immunodeficiency disorders. Clinical Immunology
(Orlando, Fla.), 160(2), 301â€“314. https://doi.org/10.1016/j.clim.2015.05.020
